### PR TITLE
Rendering PoC, start of examples - Kitchen Sink!

### PR DESCRIPTION
**Pull request by yangbin**
A terrible PoC that remakes the HTML all the time :(

but you can probably make stuff with it! go try!

Now with canvas rendering!
?renderer=jx/ui/canvas/CanvasRenderer

### CODE REVIEW

**Comment by zz85** [comment]
why does all the elements have the same id?

image

File: jx/ui/html/HTMLRenderer.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+
+HTMLRenderer.prototype.createElement = function(attr) {
+	var el = document.createElement(attr.tag || 'div');
+
+	el.style.position = 'absolute';
+
+	if (attr.text)
+		el.textContent = attr.text;
+
```
Comment (zz85) [inline jx/ui/html/HTMLRenderer.js]: you can also do
```js
if (attr.uid)
		el.id = 'jx:' + attr.uid;
```

**Comment by zz85** [comment]
Can we also rename `jx/build/JXML` -> `jx/build/JXMLBuilder` to avoid confusion?

File: jx/JXML.js [hunk 2]
```diff
@@ -79,6 +101,7 @@ JXMLComponent.prototype.create = function(module, attr) {
 		this.IDs[id] = element;
```
Comment (jieyanhuang) [inline jx/JXML.js]: where this `element` come from?

File: jx/ui/html/HTMLRenderer.js (hunk 1 above)
Comment (yangbin) [inline jx/ui/html/HTMLRenderer.js]: yep it should be in createElement not in the children loop

**Comment by yangbin** [comment]
`absolute` instead of `relative` because we are supposed to do layout ourselves

File: jx/JXML.js (hunk 2 above)
Comment (yangbin) [inline jx/JXML.js]: fixed

File: jx/ui/JXMLComponent.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+			var root = self.root = self.cloneTemplate(self.resolved.template);
+
+			var extChildren = self.attr && self.attr.externalChildren;
+
+			if (extChildren) {
+				for (var k in extChildren) {
+					extChildren['Z' + k] = extChildren[k];
+					delete extChildren[k];
+				}
+}
```
Comment (jieyanhuang) [inline jx/ui/JXMLComponent.js]: indentation spoil

**Comment by yangbin** [comment]
:haircut:

File: public/index.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	<body>
+		<script>
+			System.config({ baseURL: 'lib' });
+
+			var params = {
+				app: 'jx/example/KitchenSink',
+				renderer: 'jx/ui/html/HTMLRenderer'
+			};
+
+			location.search.replace(/\??(([^=]*)=)?([^&]+)&?/g, function(_, _, k, v) {
```
Comment (jieyanhuang) [inline public/index.html]: This regex does not seem to account for trailing `/`.
Should it be something like `/\??(([^=]*)=)?([^&]+)[^/]&?/g`?

Comment (yangbin) [inline public/index.html]: Can you give an example?

Comment (jieyanhuang) [inline public/index.html]: the problem happens when we type `?app=jx/example/KitchenSink/`, the regex
does't account for the last `/` and tries to load `KitchenSink/.jxml`.

Comment (jieyanhuang) [inline public/index.html]: the problem happens when we type `?app=jx/example/KitchenSink/`, the regex
does't account for the last `/` and tries to load `KitchenSink/.jxml`.

Comment (yangbin) [inline public/index.html]: You're not supposed to do that.

The delimiter is still open we could change it to `.` the reason being `/` is illegal in XML tags.

File: jx/build/JXML.js [hunk 5]
```diff
@@ -68,6 +68,9 @@ JXML.prototype.extractScript = function(build_assets) {
 	if (!extractScript(template))
 		build_assets.template = null;
 
+	// TODO: provide a proper scope for <script>s
+	script = 'var attr = this.jxmlcomponent.attr' + script;
```
Comment (tgohn) [inline jx/build/JXML.js]: should have a `;` after `this.jxmlcomponent.attr` to be safe.

File: jx/ui/JXMLComponent.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ * A dubious function that takes a number and
+ * returns a lexicographically sortable representation.
+ * TODO: only supports 3844 keys :(
+ *
+ * > toKey(42)
+ * "0G"
+ */
+var KEYS = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'.split(''), KEYS_LENGTH = KEYS.length;
+
+function toKey(n) {
```
Comment (tgohn) [inline jx/ui/JXMLComponent.js]: for reference only: https://github.com/deanlandolt/bytewise implements encode/decode of number, string, boolean values... for easy lexical sorting

it is overkill for our usecase anyway

**Comment by tgohn** [comment]
:+1:

File: jx/ui/JXMLComponent.js (hunk 6 above)
Comment (yangbin) [inline jx/ui/JXMLComponent.js]: Wow that's a great tool which we can use for DataNode. We've only discussed sorting for numbers (prefix with # of digits) but they've extended it to cover all types.

File: jx/build/JXML.js (hunk 5 above)
Comment (yangbin) [inline jx/build/JXML.js]: I have a commit that will fix this in a future PR

**Comment by jieyanhuang** [comment]
:+1:

File: jx/ui/JXMLComponent.js (hunk 6 above)
Comment (zz85) [inline jx/ui/JXMLComponent.js]: wow this bytewise

**Pull request merged**